wsmaster/che-core-api-workspace-activity/src/main/java/org/eclipse/che/api/workspace/activity/WorkspaceActivityChecker.java

```diff
@@ -101,19 +104,25 @@ void cleanup() {
 
   private void stopAllExpired() {
     try {
-      activityDao.findExpired(clock.millis()).forEach(this::stopExpiredQuietly);
+      activityDao
+          .findExpiredIdle(clock.millis())
+          .forEach(wsId -> stopExpiredQuietly(wsId, WORKSPACE_IDLE_TIMEOUT_EXCEEDED));
+      if (workspaceActivityManager.getRunTimeout() > 0) {
+        activityDao
+            .findExpiredRunTimeout(clock.millis(), workspaceActivityManager.getRunTimeout())
+            .forEach(wsId -> stopExpiredQuietly(wsId, WORKSPACE_RUN_TIMEOUT_EXCEEDED));
+      }
     } catch (ServerException e) {
       LOG.error("Failed to list all expired to perform stop. Cause: {}", e.getMessage(), e);
     }
   }
 
-  private void stopExpiredQuietly(String workspaceId) {
+  private void stopExpiredQuietly(String workspaceId, String reason) {
     try {
       Workspace workspace = workspaceManager.getWorkspace(workspaceId);
       workspace.getAttributes().put(WORKSPACE_STOPPED_BY, ACTIVITY_CHECKER);
       workspaceManager.updateWorkspace(workspaceId, workspace);
-      workspaceManager.stopWorkspace(
-          workspaceId, singletonMap(WORKSPACE_STOP_REASON, "Workspace idle timeout exceeded"));
+      workspaceManager.stopWorkspace(workspaceId, singletonMap(WORKSPACE_STOP_REASON, reason));
```

Comment:
@tomgeorge the reason will be logged in the stopWorkspace e.g. (Workspace run timeout exceeded), right?